### PR TITLE
Missing overrides in LoggingHttpMessageHandler and LoggingScopeHttpMessageHandler

### LINKED ISSUE #85104: Missing overrides in LoggingHttpMessageHandler and LoggingScopeHttpMessageHandler
### Background and motivation

Both [LoggingHttpMessageHandler](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs) and [LoggingScopeHttpMessageHandler](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs) have implementations of:
`protected async override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request!!, CancellationToken cancellationToken)`
but they are missing implementations of:
`protected internal override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)`
It would be great to see these implementations introduced, as otherwise custom implementation of:
- `IHttpMessageHandlerBuilderFilter`,
- `HttpHeadersLogValue`,
- `LoggingHttpMessageHandler`
- `LoggingScopeHttpMessageHandler`

is needed.
I know that synchronous usage is not expected most of the times, but we've spent quite some time figuring out why we can't turn on those logs in our legacy app migrated to .NET 6 when we tried to figure out how it works.
### Risks

_No_ response

### PR DESCRIPTION
**Pull request by mphelt**
add missing overrides in LoggingHttpMessageHandler and LoggingScopeHttpMessageHandler

Fixes https://github.com/dotnet/runtime/issues/85104

### CODE REVIEW

**Comment by mphelt** [comment]
@dotnet-policy-service agree

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs [hunk 1]
```diff
@@ -69,6 +69,24 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, CancellationTok
             }
         }
 
+        /// <inheritdoc />
+        /// <remarks>Loggs the request to and response from the sent <see cref="HttpRequestMessage"/>.</remarks>
```
Comment (Wraith2) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:73]: spelling: `Loggs` -> 'Logs`

**Comment by CarnaViire** [comment]
@mphelt thanks for your contribution!
You need to surround your overrides in `#if NET5_0_OR_GREATER` because synchronous `Send` was only added in .NET 5: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler.send
And test should have a filter as well: `[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]`

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs [hunk 2]
```diff
@@ -47,28 +47,44 @@ public LoggingHttpMessageHandler(ILogger logger, HttpClientFactoryOptions option
...
-        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        private Task<HttpResponseMessage> SendCoreAsync(HttpRequestMessage request, bool useAsync, CancellationToken cancellationToken)
         {
             ThrowHelper.ThrowIfNull(request);
             return Core(request, cancellationToken);
 
             async Task<HttpResponseMessage> Core(HttpRequestMessage request, CancellationToken cancellationToken)
             {
-                Func<string, bool> shouldRedactHeaderValue = _options?.ShouldRedactHeaderValue ?? _shouldNotRedactHeaderValue;
+                var shouldRedactHeaderValue = _options?.ShouldRedactHeaderValue ?? _shouldNotRedactHeaderValue;
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:57]:
```suggestion
                Func<string, bool> shouldRedactHeaderValue = _options?.ShouldRedactHeaderValue ?? _shouldNotRedactHeaderValue;
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs [hunk 3]
```diff
@@ -47,28 +47,44 @@ public LoggingHttpMessageHandler(ILogger logger, HttpClientFactoryOptions option
...
 
                 // Not using a scope here because we always expect this to be at the end of the pipeline, thus there's
                 // not really anything to surround.
                 Log.RequestStart(_logger, request, shouldRedactHeaderValue);
                 var stopwatch = ValueStopwatch.StartNew();
-                HttpResponseMessage response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                var response = useAsync
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:63]:
```suggestion
                HttpResponseMessage response = useAsync
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs [hunk 4]
```diff
@@ -47,28 +47,44 @@ public LoggingHttpMessageHandler(ILogger logger, HttpClientFactoryOptions option
...
                 Log.RequestEnd(_logger, response, stopwatch.GetElapsedTime(), shouldRedactHeaderValue);
 
                 return response;
             }
         }
 
+        /// <inheritdoc />
+        /// <remarks>Logs the request to and response from the sent <see cref="HttpRequestMessage"/>.</remarks>
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            => SendCoreAsync(request, true, cancellationToken);
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:79]:
```suggestion
            => SendCoreAsync(request, useAsync: true, cancellationToken);
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs [hunk 5]
```diff
@@ -47,28 +47,44 @@ public LoggingHttpMessageHandler(ILogger logger, HttpClientFactoryOptions option
...
+
+#if NET5_0_OR_GREATER
+        /// <inheritdoc />
+        /// <remarks>Logs the request to and response from the sent <see cref="HttpRequestMessage"/>.</remarks>
+        protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
+            => SendCoreAsync(request, false, cancellationToken).GetAwaiter().GetResult();
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:85]:
```suggestion
            => SendCoreAsync(request, useAsync: false, cancellationToken).GetAwaiter().GetResult();
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs [hunk 6]
```diff
@@ -58,19 +56,37 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, CancellationTok
             {
                 var stopwatch = ValueStopwatch.StartNew();
 
-                Func<string, bool> shouldRedactHeaderValue = _options?.ShouldRedactHeaderValue ?? _shouldNotRedactHeaderValue;
+                var shouldRedactHeaderValue = _options?.ShouldRedactHeaderValue ?? _shouldNotRedactHeaderValue;
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:59]:
```suggestion
                Func<string, bool> shouldRedactHeaderValue = _options?.ShouldRedactHeaderValue ?? _shouldNotRedactHeaderValue;
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs [hunk 7]
```diff
@@ -58,19 +56,37 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, CancellationTok
...
 
                 using (Log.BeginRequestPipelineScope(_logger, request))
                 {
                     Log.RequestPipelineStart(_logger, request, shouldRedactHeaderValue);
-                    HttpResponseMessage response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                    var response = useAsync
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:64]:
```suggestion
                    HttpResponseMessage response = useAsync
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs [hunk 8]
```diff
@@ -58,19 +56,37 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, CancellationTok
...
 
                     return response;
                 }
             }
         }
 
+        /// <inheritdoc />
+        /// <remarks>Logs the request to and response from the sent <see cref="HttpRequestMessage"/>.</remarks>
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            => SendCoreAsync(request, true, cancellationToken);
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:81]:
```suggestion
            => SendCoreAsync(request, useAsync: true, cancellationToken);
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs [hunk 9]
```diff
@@ -58,19 +56,37 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, CancellationTok
...
+
+#if NET5_0_OR_GREATER
+        /// <inheritdoc />
+        /// <remarks>Logs the request to and response from the sent <see cref="HttpRequestMessage"/>.</remarks>
+        protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
+            => SendCoreAsync(request, false, cancellationToken).GetAwaiter().GetResult();
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:87]:
```suggestion
            => SendCoreAsync(request, useAsync: false, cancellationToken).GetAwaiter().GetResult();
```

**Review by MihaZupan — APPROVED** [review]
Just a few comments regarding the code style used in this repo, otherwise looks good, thanks.

> Unfortunately I had to add #if NET5_0_OR_GREATER directives inside useAsync ? ... : ... - is there a cleaner way to do that?

I think it's okay as-is.

**Comment by CarnaViire** [comment]
> I've added them, but I didn't thought of it initially, as none of the base classes has those conditions - how are they different?

It's because previously only .NET Standard 2.0 API surface was used, which includes only async send: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.delegatinghandler.sendasync?view=netstandard-2.0 -- because Microsoft.Extensions.Http is compiled for .NET Standard 2.0.
But we missed the fact that when it's used in .NET 5+ apps, the sync API can be used, but it won't be overridden.

> Handlers that this PR is changing are not sealed, should I make those helper methods protected virtual?

This would be adding the helper method to the public surface, which should go through API review process: https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md
In general, I don't think it would bring too much value, as it's more of an implementation detail. Users can still derive from and substitute the logging handlers -- they would just need to override both of the Sends, which is true for deriving from DelegatingHandler as well.
As for configuring logging, there's a separate umbrella issue for that: https://github.com/dotnet/runtime/issues/77312

**Comment by mphelt** [comment]
@CarnaViire, thank you for explanation. Is there any way to re-run those failing checks? They seem to have failed for reasons not connected to this particular PR.

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks!

**Comment by CarnaViire** [comment]
Test failure is unrelated https://github.com/dotnet/runtime/issues/85772

**Pull request merged**